src/network/mining_pool.rs

```diff
@@ -0,0 +1,324 @@
+    }
+
+    /// Returns an instance of the mining pool router.
+    pub fn router(&self) -> MiningPoolRouter<N> {
+        self.mining_pool_router.clone()
+    }
+
+    /// Returns all the shares in storage.
+    pub fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.state.to_shares()
+    }
+
+    ///
+    /// Performs the given `request` to the mining pool.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: MiningPoolRequest<N>) {
+        match request {
+            MiningPoolRequest::ProposedBlock(peer_ip, mut block, miner_address) => {
+                if let Some(current_template) = &*self.current_template.read().await {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block.height() {
+                        warn!("[ProposedBlock] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Check that the block's coinbase transaction owner is the mining pool address.
+                    let records = match block.to_coinbase_transaction() {
+                        Ok(tx) => {
+                            let coinbase_records: Vec<Record<N>> = tx.to_records().collect();
+                            let valid_owner = coinbase_records
+                                .iter()
+                                .map(|r| Some(r.owner()) == self.mining_pool_address)
+                                .fold(false, |a, b| a || b);
+
+                            if !valid_owner {
+                                warn!("[ProposedBlock] Peer {} sent a candidate block with an invalid owner.", peer_ip);
+                                return;
+                            }
+
+                            coinbase_records
+                        }
+                        Err(err) => {
+                            warn!("[ProposedBlock] {}", err);
+                            return;
+                        }
+                    };
+
+                    // Determine the score to add for the miner.
+                    let proof_bytes = match block.header().proof() {
+                        Some(proof) => match proof.to_bytes_le() {
+                            Ok(bytes) => bytes,
+                            Err(err) => {
+                                warn!("[ProposedBlock] {}", err);
+                                return;
+                            }
+                        },
+                        None => {
+                            warn!("[ProposedBlock] Peer {} sent a candidate block with a missing proof.", peer_ip);
+                            return;
+                        }
+                    };
+
+                    let hash_difficulty = sha256d_to_u64(&proof_bytes);
+                    let mut info = self.miner_info.write().await;
```

Comment:
this write lock is held for too long - the related operations should happen right afterwards, and it should be released as soon as it's no longer needed